src/kudu/consensus/log_cache.cc

```diff
@@ -698,35 +572,36 @@ Status LogCache::ReadOps(int64_t after_op_index,
           next_index, up_to, remaining_space, context, &raw_replicate_ptrs),
         Substitute("Failed to read ops $0..$1", next_index, up_to));
 
+      // Compress messages read from the log if:
+      // (1) the feature is enabled through
+      // enable_compression_on_cache_miss_ flag
+      // (2) the request is not for a proxy host (the payload is discarded for
+      // a proxy request and it is wasteful to compress it here)
+      const bool should_compress =
+        enable_compression_on_cache_miss_ && !context.route_via_proxy;
+
+      vector<ReplicateMsgWrapper> msg_wrappers;
+      faststring buffer;
+
+      for (const auto& replicate : raw_replicate_ptrs) {
+        ReplicateMsgWrapper msg_wrapper(make_scoped_refptr_replicate(replicate),
+            should_compress ? codec_.load() : nullptr);
+        RETURN_NOT_OK(msg_wrapper.Init(&buffer));
+        msg_wrappers.push_back(msg_wrapper);
+      }
+
       VLOG_WITH_PREFIX_UNLOCKED(2)
-          << "Successfully read " << raw_replicate_ptrs.size() << " ops "
+          << "Successfully read " << msg_wrappers.size() << " ops "
           << "from disk (" << next_index << ".."
-          << (next_index + raw_replicate_ptrs.size() - 1) << ")";
-
-      if (enable_compression_on_cache_miss_ && !context.route_via_proxy) {
-        // Compress messages read from the log if:
-        // (1) the feature is enabled through
-        // enable_compression_on_cache_miss_ flag
-        // (2) the request is not for a proxy host (the payload is discarded for
-        // a proxy request and it is wasteful to compress it here)
-        vector<ReplicateMsg*> compressed_replicate_ptrs;
-        (void) CompressMsgs(raw_replicate_ptrs, &compressed_replicate_ptrs);
-
-        // TODO (vinay): Refactor this to not have to copy pointers again into
-        // original vector
-        raw_replicate_ptrs.clear();
-        raw_replicate_ptrs.reserve(compressed_replicate_ptrs.size());
-        raw_replicate_ptrs.insert(
-            raw_replicate_ptrs.end(),
-            compressed_replicate_ptrs.begin(),
-            compressed_replicate_ptrs.end());
-      }
+          << (next_index + msg_wrappers.size() - 1) << ")";
 
       if (!context.route_via_proxy) {
         // Compute crc checksums for the payload that was read from the log
         // Note that this is done _only_ for non-proxy requests because payload
         // is discarded for proxy requests
-        for (ReplicateMsg* msg : raw_replicate_ptrs) {
+        for (const auto& msg_wrapper : msg_wrappers) {
+          ReplicateMsg* msg = msg_wrapper.GetCompressedMsg() ?
```

Comment:
needs a comment to explain that some messages like rotates dont get compressed

Reply:
I'll add a comment but note that this is not just for rotates but also for when compression is disabled.